### PR TITLE
[ORCA-119] Implement Nextflow workflow for DCQC

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
If you want to test this workflow, you'll need to run it with the following `nextflow.config` file to use the latest (unmerged) version of py-dcqc. Once [that PR](https://github.com/Sage-Bionetworks-Workflows/py-dcqc/pull/10) is merged, I'll mint a new version of py-dcqc and update the container specified in this PR. 

### Command

```console
nextflow run Sage-Bionetworks-Workflows/nf-dcqc -r bgrande/ORCA-119/nf-dcqc -profile test_full,docker -config nextflow.config
```

### nextflow.config

```groovy
process {
    withLabel:dcqc {
        container   = 'sagebionetworks/dcqc:commit-ce57dd8'
    }
}
```

Depends on https://github.com/Sage-Bionetworks-Workflows/py-dcqc/pull/10

----

This PR is bigger than usual, but most of this code was adapted from the [dcqc-poc workflow](https://github.com/Sage-Bionetworks-Workflows/nf-dcqc-poc) that I wrote in December. In addition, the workflow effectively is a series of wrappers around `py-dcqc` commands. So, this shouldn't be too difficult to review conceptually if you've reviewed https://github.com/Sage-Bionetworks-Workflows/py-dcqc/pull/9 first. 

Note that I've initialized the `main` branch with the nf-core template, so some edits here are disabling irrelevant parts of the template. 

I was also a good Nextflow author and organized the workflow into subworkflows and modules. The process (no pun intended) was relatively straightforward. 

I might make some additional changes, but the current version is working with the `test_txt` profile (see example command and output below), so I think it's ready for a first round of review. 

```console
❯ nextflow run main.nf -profile test_txt --outdir results/
N E X T F L O W  ~  version 22.11.1-edge
Launching `main.nf` [dreamy_lamarr] DSL2 - revision: b6736347eb


------------------------------------------------------
  nf-dcqc v0.1
------------------------------------------------------
Core Nextflow options
  runName                   : dreamy_lamarr
  launchDir                 : /Users/bgrande/Repos/nf-dcqc
  workDir                   : /Users/bgrande/Repos/nf-dcqc/work
  projectDir                : /Users/bgrande/Repos/nf-dcqc
  userName                  : bgrande
  profile                   : test_txt
  configFiles               : /Users/bgrande/Repos/nf-dcqc/nextflow.config

Input/output options
  input                     : /Users/bgrande/Repos/nf-dcqc/data/input_txt.csv
  outdir                    : results/

Reference genome options
  igenomes_ignore           : true

Institutional config options
  config_profile_name       : Test TXT profile
  config_profile_description: Small test dataset to check pipeline function

Max job request options
  max_cpus                  : 2
  max_memory                : 6.GB
  max_time                  : 6.h

!! Only displaying parameters that differ from the pipeline defaults !!
------------------------------------------------------
If you use nf-dcqc for your analysis please cite:

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x

* Software dependencies
  https://github.com/nf-dcqc/blob/master/CITATIONS.md
------------------------------------------------------
executor >  local (28)
[66/dfc1e5] process > SAGE_DCQC:DCQC:PREPARE_TESTS:CREATE_TARGETS         [100%] 1 of 1 ✔
[56/79e0dc] process > SAGE_DCQC:DCQC:PREPARE_TESTS:STAGE_TARGET (0002)    [100%] 4 of 4 ✔
[35/2ebc7b] process > SAGE_DCQC:DCQC:PREPARE_TESTS:CREATE_TESTS (0002)    [100%] 4 of 4 ✔
[39/6b7a2a] process > SAGE_DCQC:DCQC:INTERNAL_TESTS:COMPUTE_TEST (0002)   [100%] 8 of 8 ✔
[ac/265ee6] process > SAGE_DCQC:DCQC:EXTERNAL_TESTS:CREATE_PROCESS (0004) [100%] 2 of 2 ✔
[a9/cb27ee] process > SAGE_DCQC:DCQC:EXTERNAL_TESTS:RUN_PROCESS (0004)    [100%] 2 of 2 ✔
[f4/5d1491] process > SAGE_DCQC:DCQC:EXTERNAL_TESTS:COMPUTE_TEST (0004)   [100%] 2 of 2 ✔
[10/e56de3] process > SAGE_DCQC:DCQC:PREPARE_REPORTS:CREATE_SUITE (0001)  [100%] 4 of 4 ✔
[1f/0137e2] process > SAGE_DCQC:DCQC:PREPARE_REPORTS:COMBINE_SUITES       [100%] 1 of 1 ✔
-[nf-dcqc] Pipeline completed successfully-
```

### CODE REVIEW

**Comment by thomasyu888** [comment]
👁️ This looks good at first glance.  I am currently running:

```
nextflow run main.nf -profile test_txt,docker --outdir results/
```

It seems to run fine - let me dedicate more time tomorrow to fully review.

File: data/input_txt.csv [hunk 1]
```diff
@@ -0,0 +1,5 @@
+url,file_type,md5_checksum
+syn://syn41864974,TXT,38b86a456d1f441008986c6f798d5ef9
```
Comment (thomasyu888) [inline data/input_txt.csv]: Lack read access to these files

Comment (BrunoGrandePhD) [reply data/input_txt.csv]: Access granted

File: modules/local/combine_suites.nf [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    label 'dcqc'
+
+    input:
+    path report_jsons
+
+    output:
+    path "suites.json"
+
+    when:
+    task.ext.when == null || task.ext.when
```
Comment (BWMac) [inline modules/local/combine_suites.nf:12]: If i am understanding this correctly, the one scenario in which these processes with this clause would not run is if `task.ext.when` is set to `false`. Where does that happen?

**Comment by thomasyu888** [comment]
Thanks.  here is the error:

```
[99/fdf7b1] process > SAGE_DCQC:DCQC:PREPARE_TEST... [100%] 1 of 1 ✔
[8d/3c88e7] process > SAGE_DCQC:DCQC:PREPARE_TEST... [100%] 4 of 4 ✔
[c2/59914a] process > SAGE_DCQC:DCQC:PREPARE_TEST... [100%] 4 of 4 ✔
[4a/d15acc] process > SAGE_DCQC:DCQC:INTERNAL_TES... [ 50%] 4 of 8, failed: 2
[e0/3b1df4] process > SAGE_DCQC:DCQC:EXTERNAL_TES... [ 50%] 1 of 2
[4c/3a2d9b] process > SAGE_DCQC:DCQC:EXTERNAL_TES... [  0%] 0 of 1
[-        ] process > SAGE_DCQC:DCQC:EXTERNAL_TES... -
[-        ] process > SAGE_DCQC:DCQC:PREPARE_REPO... -
[-        ] process > SAGE_DCQC:DCQC:PREPARE_REPO... -
Execution cancelled -- Finishing pending tasks before exit
Error executing process > 'SAGE_DCQC:DCQC:INTERNAL_TESTS:COMPUTE_TEST (0004)'

Caused by:
  Process `SAGE_DCQC:DCQC:INTERNAL_TESTS:COMPUTE_TEST (0004)` terminated with an error exit status (1)

Command executed:

  dcqc compute-test "target-0004.Md5ChecksumTest.json" "target-0004.Md5ChecksumTest.computed.json"

Command exit status:
  1

Command output:
  (empty)

Command error:
  Traceback (most recent call last):
  
    File "/usr/local/bin/dcqc", line 8, in <module>
      sys.exit(app())
               ^^^^^
  
    File "/usr/local/lib/python3.11/site-packages/dcqc/main.py", line 129, in compute_test
      test.get_status()
  
    File "/usr/local/lib/python3.11/site-packages/dcqc/tests/test_abc.py", line 66, in get_status
      self._status = self.compute_status()
                     ^^^^^^^^^^^^^^^^^^^^^
  
    File "/usr/local/lib/python3.11/site-packages/dcqc/tests/tests.py", line 30, in compute_status
      actual_md5 = self._compute_md5_checksum(file)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
    File "/usr/local/lib/python3.11/site-packages/dcqc/tests/tests.py", line 39, in _compute_md5_checksum
      with local_path.open("rb") as f:
           ^^^^^^^^^^^^^^^^^^^^^
  
    File "/usr/local/lib/python3.11/pathlib.py", line 1044, in open
      return io.open(self, mode, buffering, encoding, errors, newline)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmp7ckz1hig/no newline.txt'

Work dir:
  /Users/tyu/sage/nf-dcqc/work/dc/505d59701433c9a20e8a86e360f3b4

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line
```

**Review by thomasyu888 — CHANGES_REQUESTED** [review]
I have access, but encountered an error:
```
[99/fdf7b1] process > SAGE_DCQC:DCQC:PREPARE_TEST... [100%] 1 of 1 ✔
[8d/3c88e7] process > SAGE_DCQC:DCQC:PREPARE_TEST... [100%] 4 of 4 ✔
[c2/59914a] process > SAGE_DCQC:DCQC:PREPARE_TEST... [100%] 4 of 4 ✔
[4a/d15acc] process > SAGE_DCQC:DCQC:INTERNAL_TES... [ 50%] 4 of 8, failed: 2
[e0/3b1df4] process > SAGE_DCQC:DCQC:EXTERNAL_TES... [ 50%] 1 of 2
[4c/3a2d9b] process > SAGE_DCQC:DCQC:EXTERNAL_TES... [  0%] 0 of 1
[-        ] process > SAGE_DCQC:DCQC:EXTERNAL_TES... -
[-        ] process > SAGE_DCQC:DCQC:PREPARE_REPO... -
[-        ] process > SAGE_DCQC:DCQC:PREPARE_REPO... -
Execution cancelled -- Finishing pending tasks before exit
Error executing process > 'SAGE_DCQC:DCQC:INTERNAL_TESTS:COMPUTE_TEST (0004)'

Caused by:
  Process `SAGE_DCQC:DCQC:INTERNAL_TESTS:COMPUTE_TEST (0004)` terminated with an error exit status (1)

Command executed:

  dcqc compute-test "target-0004.Md5ChecksumTest.json" "target-0004.Md5ChecksumTest.computed.json"

Command exit status:
  1

Command output:
  (empty)

Command error:
  Traceback (most recent call last):
  
    File "/usr/local/bin/dcqc", line 8, in <module>
      sys.exit(app())
               ^^^^^
  
    File "/usr/local/lib/python3.11/site-packages/dcqc/main.py", line 129, in compute_test
      test.get_status()
  
    File "/usr/local/lib/python3.11/site-packages/dcqc/tests/test_abc.py", line 66, in get_status
      self._status = self.compute_status()
                     ^^^^^^^^^^^^^^^^^^^^^
  
    File "/usr/local/lib/python3.11/site-packages/dcqc/tests/tests.py", line 30, in compute_status
      actual_md5 = self._compute_md5_checksum(file)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
    File "/usr/local/lib/python3.11/site-packages/dcqc/tests/tests.py", line 39, in _compute_md5_checksum
      with local_path.open("rb") as f:
           ^^^^^^^^^^^^^^^^^^^^^
  
    File "/usr/local/lib/python3.11/pathlib.py", line 1044, in open
      return io.open(self, mode, buffering, encoding, errors, newline)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmp7ckz1hig/no newline.txt'

Work dir:
  /Users/tyu/sage/nf-dcqc/work/dc/505d59701433c9a20e8a86e360f3b4

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line
```

**Comment by BWMac** [comment]
I am getting the same error

File: .gitignore [hunk 3]
```diff
@@ -1,6 +1,5 @@
 .nextflow*
```
Comment (BWMac) [inline .gitignore]: When I ran the workflow, a folder called `null/pipeline_info` was created with execution reports. Should this be added to `.gitignore`?

**Review by thomasyu888 — APPROVED** [review]
This looks good to me.  I appreciate the thought that was put into behind being a "good nextflow author".  No comments about each of the tools in `local` because it essentially mimics the cli.

✈️ ✈️ ✈️

File: .gitignore (hunk 3 above)
Comment (BrunoGrandePhD) [reply .gitignore]: I updated the default `outdir` value to `results/`, which is included in the `.gitignore` file and is more understandable than `null`.

File: modules/local/combine_suites.nf (hunk 2 above)
Comment (BrunoGrandePhD) [reply modules/local/combine_suites.nf:12]: I was copying a pattern that exists with nf-core modules ([example](https://github.com/nf-core/rnaseq/blob/adce7ce9abc8f6b338b4f53d0d988ff9a0fd52ff/modules/nf-core/qualimap/rnaseq/main.nf#L18-L19)). I believe the intended usage is that the `ext` namespace can be used to set arbitrary keys like `when`. So, if you wanted to disable the `FOO` module, you could use this config (untested): 

```groovy
process {
    withName: 'FOO' {
        ext.when = false
    }
}
```

**Pull request merged**